### ISSUE TITLE
Crash when using NAs

### ISSUE THREAD

**Issue by Cevior** [cynkra/constructive#10]
# Output but error

Input
````
df <- structure(
  list(
    date = structure(
      c(18745, 18746, 18747), class = "Date"),
    `bla` = 
      c(0.069, 0.044, 0.049),
    `bla bla` = 
      c(0.046, 0.005, 0.003)
    ), 
  row.names = c(NA, 3L), 
  class = "data.frame")

construct(df)

````

Output:
````
> construct(df)
data.frame(
  date = as.Date(c("2021-04-28", "2021-04-29", "2021-04-30")),
  bla = c(0.06900000000000001, 0.044, 0.049),
  `bla bla` = c(0.046, 0.005, 0.003)
)
Error in `construct()`:
! {constructive} couldn't create code that reproduces perfectly the output
`names(original)`:  "date" "bla" "bla bla"
`names(recreated)`: "date" "bla" "bla.bla"
`original$bla bla` is a double vector (0.046, 0.005, 0.003)
`recreated$bla bla` is absent
`original$bla.bla` is absent
`recreated$bla.bla` is a double vector (0.046, 0.005, 0.003)
i use `check = FALSE` to ignore this error
Run `rlang::last_error()` to see where the error occurred.
Warning message:
Could not use colored = TRUE, as the package prettycode is not installed. Please install it if you want to see colored output or see `?print.vertical` for more information. 



> rlang::last_trace()
<error/rlang_error>
Error in `construct()`:
! {constructive} couldn't create code that reproduces perfectly the output
`names(original)`:  "date" "bla" "bla bla"
`names(recreated)`: "date" "bla" "bla.bla"
`original$bla bla` is a double vector (0.046, 0.005, 0.003)
`recreated$bla bla` is absent
`original$bla.bla` is absent
`recreated$bla.bla` is a double vector (0.046, 0.005, 0.003)
i use `check = FALSE` to ignore this error
---
Backtrace:
    x
 1. \-constructive::construct(df)
 2.   \-constructive:::check_round_trip(x, evaled, styled_code, ignore_srcref = ignore_srcref)
 3.     \-rlang::abort(...)
````

# Crash

Input
```
nadf <- structure(
  list(
    date = structure(
      c(18745, 18746, 18747), class = "Date"),
    `bla` = 
      c(0.069, 0.044, 0.049),
    `bla bla` = 
      c(0.046, 0.005, 0.003),
    `bla bla bla` =
      c(NA_real_, NA_real_, NA_real_)
    ), 
  row.names = c(NA, 3L), 
  class = "data.frame")

construct(nadf)
```

Output
```
> construct(nadf)
Error in `construct()`:
! {constructive} could not build the requested code.
Caused by error in `if (as.numeric(formatted) == x) ...`:
! missing value where TRUE/FALSE needed
Run `rlang::last_error()` to see where the error occurred.
Warning message:
In FUN(X[[i]], ...) : NAs introduced by coercion
> rlang::last_trace()
<error/rlang_error>
Error in `construct()`:
! {constructive} could not build the requested code.
Caused by error in `if (as.numeric(formatted) == x) ...`:
! missing value where TRUE/FALSE needed
---
Backtrace:
     x
  1. +-constructive::construct(nadf)
  2. | \-constructive:::try_construct(...)
  3. |   +-rlang::try_fetch(...)
  4. |   | +-base::tryCatch(...)
  5. |   | | \-base (local) tryCatchList(expr, classes, parentenv, handlers)
  6. |   | |   \-base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
  7. |   | |     \-base (local) doTryCatch(return(expr), name, parentenv, handler)
  8. |   | \-base::withCallingHandlers(...)
  9. |   \-constructive:::construct_raw(...)
 10. |     +-data_match(x, data) %||% ...
 11. |     +-constructive:::construct_idiomatic(x, data = data, ...)
 12. |     \-constructive:::construct_idiomatic.data.frame(...)
 13. |       \-constructive:::construct_apply(...)
 14. |         \-base::lapply(args, construct_raw, ...)
 15. |           \-constructive (local) FUN(X[[i]], ...)
 16. |             +-data_match(x, data) %||% ...
 17. |             +-constructive:::construct_idiomatic(x, data = data, ...)
 18. |             \-constructive:::construct_idiomatic.double(x, data = data, ...)
 19. |               +-constructive:::construct_apply(...)
 20. |               \-base::vapply(x, format_flex, character(1))
 21. |                 \-constructive (local) FUN(X[[i]], ...)
 22. \-base::.handleSimpleError(...)
 23.   \-rlang (local) h(simpleError(msg, call))
 24.     \-handlers[[1L]](cnd)
 25.       \-rlang::abort(...)
```

**Comment by moodymudskipper** [comment]
Thanks a lot!

Fixed by #11 and #12

``` r
library(constructive)

df <- structure(
  list(
    date = structure(
      c(18745, 18746, 18747), class = "Date"),
    `bla` = 
      c(0.069, 0.044, 0.049),
    `bla bla` = 
      c(0.046, 0.005, 0.003)
  ), 
  row.names = c(NA, 3L), 
  class = "data.frame")

construct(df)
#> data.frame(
#>   date = as.Date(c("2021-04-28", "2021-04-29", "2021-04-30")),
#>   bla = c(0.06900000000000001, 0.044, 0.049),
#>   `bla bla` = c(0.046, 0.005, 0.003),
#>   check.names = FALSE
#> )

nadf <- structure(
  list(
    date = structure(
      c(18745, 18746, 18747), class = "Date"),
    `bla` = 
      c(0.069, 0.044, 0.049),
    `bla bla` = 
      c(0.046, 0.005, 0.003),
    `bla bla bla` =
      c(NA_real_, NA_real_, NA_real_)
  ), 
  row.names = c(NA, 3L), 
  class = "data.frame")

construct(nadf)
#> data.frame(
#>   date = as.Date(c("2021-04-28", "2021-04-29", "2021-04-30")),
#>   bla = c(0.06900000000000001, 0.044, 0.049),
#>   `bla bla` = c(0.046, 0.005, 0.003),
#>   `bla bla bla` = c(NA_real_, NA_real_, NA_real_),
#>   check.names = FALSE
#> )
```

<sup>Created on 2022-08-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by moodymudskipper** [comment]
Yet a bit better with #14 regarding the display of `0.69`

```
library(constructive)

df <- structure(
  list(
    date = structure(
      c(18745, 18746, 18747), class = "Date"),
    `bla` = 
      c(0.069, 0.044, 0.049),
    `bla bla` = 
      c(0.046, 0.005, 0.003)
  ), 
  row.names = c(NA, 3L), 
  class = "data.frame")

construct(df)
#> data.frame(
#>   date = as.Date(c("2021-04-28", "2021-04-29", "2021-04-30")),
#>   bla = c(0.069, 0.044, 0.049),
#>   `bla bla` = c(0.046, 0.005, 0.003),
#>   check.names = FALSE
#> )

nadf <- structure(
  list(
    date = structure(
      c(18745, 18746, 18747), class = "Date"),
    `bla` = 
      c(0.069, 0.044, 0.049),
    `bla bla` = 
      c(0.046, 0.005, 0.003),
    `bla bla bla` =
      c(NA_real_, NA_real_, NA_real_)
  ), 
  row.names = c(NA, 3L), 
  class = "data.frame")

construct(nadf)
#> data.frame(
#>   date = as.Date(c("2021-04-28", "2021-04-29", "2021-04-30")),
#>   bla = c(0.069, 0.044, 0.049),
#>   `bla bla` = c(0.046, 0.005, 0.003),
#>   `bla bla bla` = c(NA_real_, NA_real_, NA_real_),
#>   check.names = FALSE
#> )


```
